tardis/plasma/properties/ion_population.py

```diff
@@ -239,29 +241,198 @@ def calculate(self, phi, partition_function, number_density):
                               / n_electron < n_e_convergence_threshold):
                 break
             n_electron = 0.5 * (new_n_electron + n_electron)
-        return ion_number_density
+        return ion_number_density, n_electron
 
-class ElectronDensity(ProcessingPlasmaProperty):
+class IonNumberDensityNLTE(ProcessingPlasmaProperty):
     """
-    Calculate the electron density using the Saha equation with the ion
-    population ratio of a particular element.
+    Calculate the ionization balance
+
+    .. math::
+        N(X) = N_1 + N_2 + N_3 + \\dots
+        N(X) = (N_2/N_1) \\times N_1 + (N3/N2) \\times (N_2/N_1) \\times N_1 + \\dots
+        N(X) = N_1(1 + N_2/N_1 + (N_3/N_2) \\times (N_2/N_1) + \\dots
+        N(X) = N_1(1+ \\Phi_{i,j}/N_e + \\Phi_{i, j}/N_e \\times \\Phi_{i, j+1}/N_e + \\dots)
+
     """
-    latex_formula = r'$n_e = \frac{N_{i,j}}{N_{i,j+1}} \times \Phi_{i,j}$'
-
-    name = 'electron_densities'
-
-    def calculate(self, ion_number_density, phi):
-        # Could check electron density for any element/ions in each zone, but
-        # if number density of element/ions was zero, it would not work.
-        # So setting this to calculate from the dominant ion in each zone.
-        dominant_ions = ion_number_density.sum(level=(0, 1)).idxmax()
-        upper_ions = []
-        for ion in dominant_ions.values:
-            upper_ions.append((ion[0], ion[1] + 1))
-        n_electron = []
-        for zone in range(len(ion_number_density.columns)):
-            n_electron.append(
-                (ion_number_density[zone].ix[dominant_ions[zone]] /
-                 ion_number_density[zone].ix[upper_ions[zone]]) *
-                phi.ix[upper_ions[zone]][zone])
-        return pd.Series(n_electron, index=np.arange(0, len(n_electron)))
\ No newline at end of file
+
+    latex_formula = (r'$N(X) = N_1 + N_2 + N_3 + \dots \\ '
+                     r'N(X) = (N_2/N_1) \times N_1 + (N3/N2) '
+                     r'\times (N_2/N_1) \times N_1 + \dots \\'
+                     r'N(X) = N_1(1 + N_2/N_1 + (N_3/N_2) \times (N_2/N_1) '
+                     r'+ \dots \\'
+                     r'N(X) = N_1(1+ \Phi_{i,j}/N_e + \Phi_{i, j}/N_e '
+                     r'\times \Phi_{i, j+1}/N_e + \dots)$')
+
+    outputs = ('ion_number_density', 'electron_densities',
+               'level_boltzmann_factor')
+
+    def __init__(self, plasma_parent, ion_zero_threshold=1e-20):
+        super(IonNumberDensityNLTE, self).__init__(plasma_parent)
+        self.ion_zero_threshold = ion_zero_threshold
+
+    def calculate_nlte(self, nlte_input, beta_sobolev, j_blues, levels,
```

Comment:
There's a loop of property dependencies that goes: 
`level_boltzmann_factor` (from nlte level population calculation) -> `partition_function` -> `ion_number_density` -> `level_number_density` -> `tau_sobolevs` -> `beta_sobolevs` -> `level_boltzmann_factor` (nlte level population calculation again). So I thought we needed that loop to be inside one plasma property for consistency.